### ISSUE TITLE
optional and shared server(s) storage paths

### ISSUE THREAD

**Issue by jxsl13** [teeworlds/teeworlds#2254]
### Basic problem: 
A server executable currently searches for file relative to the executable's position.
This makes server config files rather ugly with all of their `../../`.
Especially when someone runs multiple servers at once and wants to share configurations with "exec shared.cfg" or share maps in a convenient manner.

### Proposal 1: 
Changable maps folder location, in order to share maps between multiple server instances.
I'm not sure how to best implement this, as I'm not familiar with the storage mechanics.
I would assume that passing this as a command line argument would make sense and as command line arguments are actually parsed as rcon commands, it would maybe make also sense to implement that command.

### Proposal 2:
As config files etc. are looked for relative to the executable, it might make the server administration a lot easier(especially with lots of servers), if one could provide a directory which is instead used as the base directory where the files are being looked for relatively to.(broken english, sry.)

Is any of the previous options set, the the default behaviour should be omitted and nothing should be looked for relative to the executable.
I don't know whether the storage.cfg is actually being used by the server, but if yes, there could be two extra variables specified where to look for maps and where to look for any path that is being referenced, like "exec" configs or log files, to name two(maybe these two are actually all of the commands that use this. Other commands that create directories might create those in the specified *directory*).

Here is an example of one such configuration, tell me what's going on assuming the executable lies in `~/zcatch/build/`:D
<img width="1446" alt="Bildschirmfoto 2019-10-01 um 01 29 44" src="https://user-images.githubusercontent.com/371944/65924052-5686b800-e3eb-11e9-8ab6-4cfcbdf3eb72.png">

Edit:
The `-f` command line argument does not seem to allow relative paths at all.(paths that do a kinda directory traversal into a parent directory)

Edit 2:
As heinrich5991 hinted at, It seems that `exec` does not allow to traverse to parent directories, most likely due to security reasons, which makes shared configurations between different server modifications a hassle?
Basically one would need to put all of their executables in one folder, one big `/data/maps/`folder and live without any server mod updates, because those would need different git repositories, which cannot be put into one directory like the executables could be.

Edit 3:

One could maybe use the initially passed config file(`-f ../../allow/traversal/server001.cfg`) as the relative path from which everything is defined, so one could have the individual configs at the same level. then some directories for e.g. logs, shared configs, maps at the same level as the config files.
At runtime the `exec` settings the should not be able to traverse to parent directories.

Edit 4:
I think that a lot of servers are being run with root privileges and I wonder, why the `logfile ../../logs/Whatever-`is actually able to traverse to a parent directory.. someone with full rcon access might redirect the logfile to be written into some system relevant config files and hijjack the whole vps/server(I dunno how, but I see a possibility).

**TLDR:**

Best approach would be to have the storage system look relative to the path of the passed configuration file `teeworlds_srv -f ~/server_configs/lms.cfg`. In this specific case it should not matter, where the config file lies. It should only matter when executing other config files after the first one has been executed and restrict further `exec` commands from traversing to parent directories.
Another approach should be thought up for shared maps, maybe a command line argument like `-f`

**Comment by oy** [comment]
Use the `storage.cfg` that's supposed to be used for that.

**Comment by jxsl13** [comment]
I don't seem to be able to get it properly to work, is there some example, where I can define a directory for configs and one for maps and hierarchically configure the servers in a top-down manner?

**Comment by Dune-jr** [comment]
It it my understand that you can add paths in https://github.com/teeworlds/teeworlds/blob/master/storage.cfg as such:

```
add_path ~/tw/configs
add_path ~/tw/maps
```

> hierarchically configure the servers in a top-down manner?

sorry?

**Comment by jxsl13** [comment]
that sadly didn't work, as I said out in the discord :/
At least not with the 

With the second part I mean to have for example your your top level configurations for your different servers
```
~/configs/server1.cfg
~/configs/server2.cfg
~/configs/shared/rcon.cfg // included in server1.cfg, server2.cfg
~/configs/maps/maps_mod1.cfg
~/config/maps/maps_mod2.cfg
```

preferable one can access configs in a maps folder directly, as for example specific map repositories can contain an own maps.cfg file where all the maps are automatically added.


> seems like combination of `add_path ~/` or `add_path ~` in combination with `teeworlds_srv -f config/config.cfg` or` teeworlds_srv -f ~/config/config.cfg` doesn't work,

Also mentioned, that the storage.cfg isn't copied into the make bold folder.

**Comment by oy** [comment]
> ~/configs/server1.cfg
~/configs/server2.cfg
~/configs/shared/rcon.cfg // included in server1.cfg, server2.cfg
~/configs/maps/maps_mod1.cfg
~/config/maps/maps_mod2.cfg

preferable one can access configs in a maps folder directly, as for example specific map repositories can contain an own maps.cfg file where all the maps are automatically added.

Add the line `add_path ~/configs` to your storage.cfg and then you can f.e. do `exec server1.cfg` in rcon.

**Comment by jxsl13** [comment]
I actually would like to structure server config files in a maintainable manner.
I don't want to execute configs from the rcon, but on startup from within other server configs.

![Bildschirmfoto 2019-10-02 um 00 04 20](https://user-images.githubusercontent.com/371944/66004682-ac219a00-e4a9-11e9-8324-6b854bc48381.png)

### 1. discovery
It seems like relative paths like `add_path ~/servers` are not supported.
Instead when adding the full path `add_path /home/user/servers`, then multiple directories are created in that added folder.
them being: `configs`, `demos`, `dumps`

### 2. discovery
It seems that maps are being properly found in these directories, but log files don't seem to follow any such schema. Sadly one cannot create log files in the same manner as one can access the config files via `configs/shared/rcon.cfg`.

### 3. discovery
after a while, the server creates a `masters.cfg` file.

<img width="947" alt="Bildschirmfoto 2019-10-02 um 00 50 04" src="https://user-images.githubusercontent.com/371944/66006848-6a482200-e4b0-11e9-861a-5f25d3070c2c.png">
(green arrows: works fine, red arrow: might hopefully be added and made to work)

I think that it might make sense to add log files to this storage functionality and maybe support home directory paths `~/` or `~`.

**Comment by cClaude** [comment]
> I don't seem to be able to get it properly to work, is there some example, where I can define a directory for configs and one for maps and hierarchically configure the servers in a top-down manner?

What about using links ?
```bash
ln -s /path/global/data /path/server_n/data
```
or

```bash
ln -s /path/global/data/maps /path/server_n/data/maps
```

**Comment by ChillerDragon** [comment]
Meh https://github.com/teeworlds/teeworlds/commit/c705f048f3f62e0ed92686e19763a61309125d98 killed all my setups :/
Is there any hope for supporting absolute paths again? Maybe if the path starts with a '/' ignore the dumps/ directory and just use absolute. Fiddling around with links is confusing imo. I know absolute paths can be a security risk but admins who start their server with the right permissions won't get their ``/etc/shadow`` overwritten hehe. Maybe just support absolute paths config only so in game admins can't mess it up.

**Comment by heinrich5991** [comment]
Make a link `root` pointing to `/` in every storage directory and then run a `sed` replacing `exec /` with `exec root/`.

**Comment by ChillerDragon** [comment]
Yikes @heinrich5991
Well I somehow manged to get my system running again but, same as your suggestion, it was effort and a lil confusing. And do we really want to increase the effort to update to 0.7?

**Comment by heinrich5991** [comment]
Getting from a game server to arbitrary file read on a is a security issue in my eyes, so I agree that we shouldn't support that.

However, the error messages should be improved, to only trigger the "absolute paths" warning when an absolute path is given; perhaps even erroring out completely without the confusing "file not found" error.
